plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/inspections/logging/KotlinPlaceholderCountMatchesArgumentCountInspection.kt

```diff
@@ -0,0 +1,208 @@
+import org.jetbrains.kotlin.psi.KtValueArgument
+import org.jetbrains.kotlin.psi.callExpressionVisitor
+import org.jetbrains.kotlin.resolve.calls.callUtil.getType
+import org.jetbrains.kotlin.resolve.constants.evaluate.ConstantExpressionEvaluator
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+import org.jetbrains.kotlin.resolve.lazy.BodyResolveMode
+import org.jetbrains.kotlin.types.KotlinType
+import org.jetbrains.kotlin.types.typeUtil.supertypes
+import org.jetbrains.kotlin.utils.addToStdlib.safeAs
+
+class KotlinPlaceholderCountMatchesArgumentCountInspection : AbstractKotlinInspection() {
+
+    enum class LoggerType {
+        SLF4J_LOGGER, LOG4J_LOGGER, LOG4J_BUILDER
+    }
+
+    companion object {
+        private val LOGGING_METHODS = setOf("log", "trace", "debug", "info", "warn", "error", "fatal")
+        private val LOGGER_CLASSES =
+            mapOf(
+                "org.slf4j.Logger" to LoggerType.SLF4J_LOGGER,
+                "org.apache.logging.log4j.Logger" to LoggerType.LOG4J_LOGGER,
+                "org.apache.logging.log4j.LogBuilder" to LoggerType.LOG4J_BUILDER
+            )
+    }
+
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean) = callExpressionVisitor(fun(call) {
+        val name = call.calleeExpression?.text
+        if (name == null || !LOGGING_METHODS.contains(name)) {
+            return
+        }
+
+        val method = call.resolveToCall()?.resultingDescriptor ?: return
+        val parents = method.receiverType().findParents(LOGGER_CLASSES.keys)
+        if (parents.isEmpty()) {
+            return
+        }
+        val loggerType = LOGGER_CLASSES[parents.first()] ?: return
+
+        val valueParameters = method.valueParameters
+        if (valueParameters.isEmpty()) {
+            return
+        }
+        val indexString: Int
+        if (KotlinBuiltIns.isString(valueParameters[0]?.type)) {
+            indexString = 0
+        } else {
+            if (valueParameters.size < 2 || !KotlinBuiltIns.isString(valueParameters[1]?.type)) {
+                return
+            }
+            indexString = 1
+        }
+
+        val valueArguments = call.valueArguments
+        val patternArgument = valueArguments[indexString] ?: return
+        val pattern = patternArgument.getArgumentExpression() ?: return
+
+        var argumentsCount = valueArguments.size - 1 - indexString
+        var lastArgumentIsException = valueArguments.lastOrNull()?.getArgumentExpression()?.resolveType().isThrowable()
+
+        if (argumentsCount == 1) {
+            val ktValueOneArgument = valueArguments[indexString + 1] ?: return
+            val oneArgument = ktValueOneArgument.getArgumentExpression()
+            val resolveType = oneArgument?.resolveType()
+            if (resolveType != null && KotlinBuiltIns.isArray(resolveType) && ktValueOneArgument.isSpread) {
+                val callExpression = oneArgument.safeAs<KtCallExpression>() ?: return
```

Comment:
It's unclear why argument is checked against expression only. There are other cases (reference, function call) and other ways to create an array. This should be either covered or explained (why not).

Reply:
This case only for logger.debug("test {} {}", *arrayOf(1, 2, 3)), because sometimes it appears after auto-conversion from java. Other cases are possible, of course, but I am not sure, that they are widespread. In Java inspection they are not considered too. Add some comments

Reply:
thank you